src/cmap/establish/test.rs

```diff
@@ -0,0 +1,155 @@
+        Some(&AuthMechanism::MongoDbX509) | None => Vec::new(),
+        Some(other) => vec![other.clone()],
+    };
+
+    client
+        .drop_and_create_user(
+            credential.username.as_deref().unwrap(),
+            credential.password.as_deref(),
+            &[role.into()],
+            &mechanisms,
+            credential.source.as_deref(),
+        )
+        .await
+        .unwrap();
+
+    let pool_options = ConnectionPoolOptions::builder()
+        .credential(credential.clone())
+        .tls_options(CLIENT_OPTIONS.tls_options())
+        .build();
+
+    let handshaker = Handshaker::new(Some(&pool_options));
+
+    let mut conn = Connection::new(1, Default::default(), 1, Some(pool_options.into()))
+        .await
+        .unwrap();
+
+    let first_round = handshaker
+        .handshake(&mut conn, Some(&credential))
+        .await
+        .unwrap();
+
+    // We expect that the server will return a response with the `speculativeAuthenticate` field if
+    // and only if it's new enough to support it.
+    assert_eq!(first_round.is_some(), client.server_version_gte(4, 4));
+
+    // Regardless of whether the server supports our speculative authentication attempt, we should
+    // be able to successfully authenticate after the handshake.
+    credential
+        .authenticate_stream(&mut conn, &Default::default(), first_round)
+        .await
+        .unwrap();
+
+    let command = Command::new(
+        "find".into(),
+        authorized_db_name.into(),
+        doc! { "find": "foo", "limit": 1  },
+    );
+
+    let response = conn.send_command(command, None).await.unwrap();
+
+    assert!(response.is_success());
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn speculative_auth_default() {
+    let _guard = LOCK.run_concurrently().await;
+
+    let client = TestClient::new().await;
+
+    let credential = Credential::builder()
+        .username(function_name!().to_string())
+        .password("12345".to_string())
+        .build();
+
+    speculative_auth_test(&client, credential, "root", function_name!()).await;
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn speculative_auth_scram_sha_1() {
+    let _guard = LOCK.run_concurrently().await;
+
+    let client = TestClient::new().await;
+
+    let credential = Credential::builder()
+        .username(function_name!().to_string())
+        .password("12345".to_string())
+        .mechanism(AuthMechanism::ScramSha1)
+        .build();
+
+    speculative_auth_test(&client, credential, "root", function_name!()).await;
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn speculative_auth_scram_sha_256() {
+    let _guard = LOCK.run_concurrently().await;
```

Comment:
I realized that the X.509 test was the only one that conflicted with anything; the SCRAM tests all use different users, so I've switched them to run concurrently